mlir/include/mlir/IR/CommonAttrConstraints.td

```diff
@@ -408,17 +408,26 @@ class ElementsAttrBase<Pred condition, string summary> :
   let storageType = [{ ::mlir::ElementsAttr }];
   let returnType = [{ ::mlir::ElementsAttr }];
   let convertFromStorage = "$_self";
+
+  // The underlying C++ value type of each element.
+  string elementReturnType = ?;
```

Comment:
This may need wider support, specifically we may need to generate an accessor function in .h.inc/.cpp.inc.  Something like https://github.com/llvm/llvm-project/blob/main/mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp#L1204, for example.

I'm wary about making this kind of change in a widely shared file.  Maybe we could just handle this in OmpOpGen.cpp?  Specifically, infer this information in there based on the type of the attribute?

Reply:
> I'm wary about making this kind of change in a widely shared file. Maybe we could just handle this in OmpOpGen.cpp? Specifically, infer this information in there based on the type of the attribute?

Yes, this is something I tried to avoid as well. The problem is that the only existing attribute we could potentially use to get the element type information is the `returnType` inherited from `Attr`. We could potentially remove the "::llvm::ArrayRef<>" part of that string in the case of `DenseArrayAttrBase` and derived types, which doesn't seem like a very clean solution but it would work (as long as these subclasses/definitions don't override that property). For other subclasses of `ElementsAttrBase` we would have to accept having to use array-style attributes (e.g. `::mlir::DenseIntElementsAttr`) instead of lists of elements.

I'd like to avoid hardcoding as many type names as possible in the new tablegen backend, since people could just create new general or OpenMP-specific attribute types and then it would have to be updated. I think it makes sense to specialize it for as few and as generic cases as we can get away with and just make sure they already contain the information we need. In this case, we're just missing the element type and rank of array attributes, which seems like something that could be of general use eventually.

Having said that, this is just the approach that works that made the most sense to me, but I'm very much interested in discussing potentially better alternatives.

> This may need wider support, specifically we may need to generate an accessor function in .h.inc/.cpp.inc.

Good point, I'll delay making this change until we decide whether we want to keep these new properties or not.

Reply:
Tagging people involved in the patch that created the CommonAttrConstraints.td file and reviewers (sorry for the spam, just looking for an expert in this area to comment on these potential additions!): @Groverkss, @River707, @Mogball, @joker-eph.

Reply:
Yeah I would consider it a bit off to make changes to the ODS spec only for OMP purposes. What exactly are you trying to do here?

Reply:
Thank you @Mogball for taking a look.

> What exactly are you trying to do here?

The main thing for which I added these properties is to be able to translate argument types as defined in tablegen to some corresponding C++ types used to store this information in a way that is useful for us. For example:
- `OptionalAttr<I32Attr>` -> `::mlir::IntegerAttr`
- `DeclareTargetAttr` -> `::mlir::omp::DeclareTargetAttr`
- `DenseI8ArrayAttr` -> `::llvm::SmallVector<int8_t>`
- `I32` -> `::mlir::Value`
- `Variadic<Index>` -> `::llvm::SmallVector<::mlir::Value>`

For values, this is not an issue. For attributes, we can generally use already existing properties, such as their `storageType`. The issue comes when trying to represent definitions based on `ElementsAttrBase`, like `DenseI8ArrayAttr` above, because we don't have a general way of knowing what element type we should use for the generated vector. That's what the `elementReturnType` (name can most likely be improved) solves, it just holds the `cppType` template parameter of `DenseArrayAttrBase` definitions and provides some hopefully reasonable types for other tablegen subclasses, so that it can be queried independently.

Then, it made sense to me that we should be able to also have access to the rank of these `ElementsAttrBase` definitions and generate some specific code for 2+D arrays, but right now it's not for certain that we would actually need that, so I'm thinking support for this could be postponed to when/if it's needed.

Reply:
Hello again @Mogball, have you had a chance to think about this proposal? I'd like to decide on an approach to unblock progress on this work. Thank you again!

Reply:
Ping again! Is there anything I can do to unblock progress on this? Thank you for your time.

Reply:
Sorry, I don't know why my email notifications aren't going through. I got this one now

Reply:
> in a way that is useful for us

This is kind of the crux of the issue. There is a kind of layering inversion here: only the OMP tablegen backend makes use of this information, yet the onus is on the main tablegen definitions to provide this information. No other dialects will realistically ever have to provide these definitions, and it's not clear to them why these are useful if they aren't using OMP.

I don't think there's a way forward for adding this kind of metadata to ODS proper if there is no conceivable use outside of OMP. Since the information has to be specified manually for ODS definitions to be used with OMP, would it be much worse to have a centralized mapping of `Attr -> C++ representation`?

Reply:
Thank you for the discussion.

> Since the information has to be specified manually for ODS definitions to be used with OMP, would it be much worse to have a centralized mapping of Attr -> C++ representation in the OMP backend?

I will look into creating such a mapping, though my concern about that approach is that it's unlikely to be very general. That is, I think it might be easy to break it by just using some new attribute definition based on `ElementsAttrBase`. I think the addition of a field to hold the element type would be more resilient in this regard, since it would always be present regardless of levels of inheritance, etc (except if explicitly overridden to prevent it).

> There is a kind of layering inversion here: only the OMP tablegen backend makes use of this information, yet the onus is on the main tablegen definitions to provide this information. No other dialects will realistically ever have to provide these definitions, and it's not clear to them why these are useful if they aren't using OMP.
>
> I don't think there's a way forward for adding this kind of metadata to ODS proper if there is no conceivable use outside of OMP.

I share that concern, but at the same time I feel like what I'm proposing of adding is a property to hold a subset of data that is already defined by `DenseArrayAttrBase` (the `cppType` template argument). It's adding some flexibility to query that information that initially would only be used by the OpenMP dialect but which is not really specific to it. We don't have any uses for it on other dialects at this point, but I think there is no reason why it couldn't have other use cases later on.

Also, apart from subclasses/definitions based directly on `ElementsAttrBase` (of which I can currently see none outside of CommonAttrConstraints.td), no other attribute types defined in a dialect would actually have to initialize that new property, as this is done for all direct subclasses/definitions in this PR. So, in that sense, it would be a transparent addition.

Reply:
I just added a commit to undo additions to base MLIR attribute definitions and to update the implementation of the tablegen backend to get the element type information from existing information (hard-coding special handling for a couple more attribute types). Let me know if this approach addresses your concerns and is something we could consider merging @Mogball, @kparzysz, @Meinersbur. Thank you!